design/open/spark-outputcommitter/committer.md

```diff
@@ -0,0 +1,194 @@
+prepare the desired output, and merge back in as part of the HOC commit.[^1]
+Aborting will be done by dropping the branch (or repurposing it if the same
+job ID is requested again).  Regular lakeFS retention can handle dropping
+data objects; we might want to add file patterns to retention to allow
+temporary objects to be dropped rapidly.
+
+[^1]: A merge is a type of commit, of course, so in this model HOC commits
+	*are* lakeFS commits!
+
+# How
+
+## Conflict modes
+
+Spark supports multiple "save modes": Append, Overwrite, ErrorIfExists, and
+Ignore.  These impact conflict resolution.  We will initially support just
+overwrite modes: an entire previous table will be deleted on write.
+
+## Sample flow
+
+* User configures lakeFSFS and configures LakeFSOutputCommitter as the
+  OutputCommitter for `lakefs` protocol Paths.  **Possibly** lakeFSFS will
+  set up this OutputCommitter by default.
+
+  Successive steps are controlled by Spark/Hadoop to output, and correspond
+  to the Hadoop OutputCommitter protocol.
+* **Setup (job/task TBD)**: Create a new branch for this job/task.  Its name
+  is predictable from the job ID and/or task ID, so can be easily found
+  again.  In "overwrite" mode, immediately delete the entire subtree of the
+  intended output path, and possibly delete branches of previous tasks.
+  This handles cases where the names of the objects written by the output
+  format change, for instance because of repartitioning to a smaller (or
+  different) number of partitions or because of nondeterminism in the names
+  of the objects.
+* **Write objects**: Everything is written to its correct path on the branch.
+* **[HOC] Commit**: Merge back to the original branch.
+
+## Properties
+
+* The merge is performed by lakeFS so it is **atomic**.
+* In the **single-writer case** the merge succeeds: no other operations
+  occur on the subtree.
+* **In-place updates** work: the old objects are deleted and replaced by new
+  objects.  This is true regardess of partitioning etc.
+* **Multiple writers** are detected and the first fails.  The first to
+  HOC-commit succeeds.  But all the others fail: they deleted the same
+  previous files, or created a conflicting file (at least their `_SUCCESS`
+  indicator).  So their lakeFS merge fails due to a conflict with the first
+  (successful) merge.
+* (Conflicting) **non-OutputCommitter writes are detected** and clearly
+  handled.  As long as other writes create _one_ object with an overlapping
+  name the merge will fail.  So LakeFSOutputCommitter can achieve its
+  correct semantics regardless of other writers used.
+* **Clearly correct by construction**: Rather than rely on single atomic
+  operations and carefully tailoring operations to Spark retry mechanisms,
+  we use lakeFS capabilities and guarantees.  Analyzing correctness becomes
+  simpler.
+* **Fast**: No data copies, just only (required) metadata operations.  Cost
```

Comment:
When you say "no data copies" here you mean that the lakeFS operations create branch and merge don't copy data or it implies that the new HOC relies on non-copying rename?